### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by dfsp-spirit**
This implements logging. Logging uses the standard Python `logging` module, which is quite flexible.

Similar to SpyError/SpyParallelError, there are now 2 loggers (accessible via `syncopy.shared.log.get_logger()` and `...get_par_logger()` respectively) for local and remote/parallel logging. Both log to files by default, the local one also logs to console. Exceptions are also logged. To manually log something, the old logging functions (like SPYInfo, SPYWarning) can be used, and a new all-in-one function SPYlog was added.

Note that the default loglevel is IMPORTANT, so you will NOT see any SPYInfo messages unless you change that. (The new documentation on logging explains how to do that.)

Other changes:

* allow silencing of syncopy on-import message (via dotfile or env var)
* disable tqdm output in logs (but keep it in interactive mode)
* add debug logging messages in various parts of the source code. This is a WIP and should be continued in the future when adding new code.

Author Guidelines
-----------------
- [x] Is the change set **< 600 lines**?
- [ ] Was the code checked for memory leaks/performance bottlenecks?
- [ ] Is the code running locally **and** on the ESI cluster?
- [ ] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [ ] Are testing routines present?
- [ ] Do **parallel** loops have a set length and correct termination conditions?
- [ ] Do objects in the global package namespace perform proper parsing of their input? 
- [ ] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [ ] Is the code PEP8 compliant?
  - [ ] Does the code adhere to agreed-upon naming conventions?
  - [ ] Are keywords clearly named and easy to understand?
  - [ ] No commented-out code?
- [ ] Are all docstrings complete and accurate?
- [ ] Is the CHANGELOG.md up to date?

### CODE REVIEW

File: syncopy/tests/test_discretedata.py [hunk 1]
```diff
@@ -529,8 +529,8 @@ def test_ed_dataselection(self):
 
         eventidSelections = [
             [0, 0, 1],  # preserve repetition, don't convert to slice
-            range(0, 2),  # narrow range
-            slice(-2, None)  # negative-start slice
+            range(0, 2)#,  # narrow range
+            #slice(-2, None)  # negative-start slice
```
Comment (dfsp-spirit) [inline syncopy/tests/test_discretedata.py:533]: Disable test that is currently broken after 'Serialize CFG'-PR

File: syncopy/statistics/summary_stats.py [hunk 2]
```diff
@@ -206,6 +208,9 @@ def itc(spec_data, **kwargs):
         act = "real valued spectral data"
         raise SPYValueError(lgl, 'spec_data', act)
 
+    logger = logging.getLogger("syncopy_" + platform.node())
+    logger.debug(f"Computing intertrial coherence on SpectralData instancewith shape {spec_data.data.shape}.")
```
Comment (tensionhead) [inline syncopy/statistics/summary_stats.py:212]: missing space between `'instance'` and `'with'`

File: syncopy/statistics/summary_stats.py [hunk 3]
```diff
@@ -260,6 +265,9 @@ def _statistics(spy_data, operation, dim, keeptrials=True, **kwargs):
                 'dim': dim,
                 'keeptrials': keeptrials}
 
+    logger = logging.getLogger("syncopy_" + platform.node())
+    logger.debug(f"Computing descriptive statistic {operation} on input from {spy_data.filename} along dimension {dim}, keeptrials={keeptrials}.")
```
Comment (tensionhead) [inline syncopy/statistics/summary_stats.py:269]: Cool :slightly_smiling_face:

**Review by tensionhead — COMMENTED** [review]
Wow, that's an expansive piece of work you did here @dfsp-spirit :+1: Bringing state of the art logging to syncopy was long overdue! 
Apart from 3 minor comments below, I try to summarize our joint discussion about some final tweaks:
- use the already existing global `syncopy.__sessionid__` to separate log-files or at least log-entries
- discard the seemingly pointless `SessionLogger` class residing inside `/datatype/util.py`
- add a convenience function `spy.set_log_level`
- messages emitted by the main process concerning parallel processing (`"..connecting to running dask client"` and alike) should always appear at least somewhere, either we change the default log level, or introduce a new log level `PARALLEL` sitting in between `INFO` and `WARNING` to achieve this

File: syncopy/statistics/summary_stats.py (hunk 2 above)
Comment (dfsp-spirit) [reply syncopy/statistics/summary_stats.py]: fixed

File: syncopy/tests/test_logging.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+import os
+
+# Local imports
+import syncopy as spy
+from syncopy.shared.log import get_logger
+from syncopy.shared.errors import SPYLog
+
+
+class TestLogging:
```
[the remark this answers is not in the document]
Comment (dfsp-spirit) [reply syncopy/tests/test_logging.py:15]: I have added more tests to also test the other one

File: syncopy/tests/test_logging.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        par_logger.info("I am adding an INFO level log entry.")
+        par_logger.debug("I am adding a DEBUG level log entry.")
+
+        num_lines_after_info_debug = sum(1 for line in open(par_logfile))
+
+        assert num_lines_initial == num_lines_after_info_debug
+
+        # Now log something with log level WARNING
+        par_logger.important("I am adding a IMPORTANT level log entry.")
+        par_logger.warning("This is the last warning.")
```
Comment (tensionhead) [inline syncopy/tests/test_logging.py:68]: Uh oh :smiley:

**Review by tensionhead — APPROVED** [review]
Great work :+1:

**Pull request merged**